### ISSUE TITLE
handling log(cosh(x)) for large x

### ISSUE THREAD

**Issue by laolux** [sympy/sympy#12671]
I am trying to plot something containing log(cosh(x)) with large x which sympy somehow does plot, but it tells me to report a bug.
/usr/lib/python3.5/site-packages/sympy/plotting/experimental_lambdify.py:230: UserWarning: The evaluation of the expression is problematic. We are trying a failback method that may still work. Please report this as a bug.

The problem with the resulting plot is that it is not exactly symmetric even though it should be.
I suppose the problem is that cosh(1000) is very large, but taking the logarithm afterwards yields something reasonable, log(cosh(1000)) \approx 1000.

I have included a minimal working example.
In the example you will also see a slightly modified version: sign(x)*x + log(exp(-sign(x)*2*x)+1) - log(2).

Mathematically those expressions are identical for real x, but sympy does not notice that and thus does not properly simplify the expression.
In the example one can also see a difference plot for those equivalent expressions. Please note that this plot is not deterministic, i.e. it changes every execution.
For comparison I did include a numpy plot, especially to show the behaviour for small x.

MWE:

    import sympy as sym
    
    class log_cosh(sym.Function):
        nargs=1
    
        @classmethod
        def eval(cls, x):
            return sym.log(sym.cosh(x))
    
    class replacement(sym.Function):
        nargs=1
        
        @classmethod
        def eval(cls, x):
            return sym.sign(x)*x + sym.log(sym.exp(-sym.sign(x)*2*x)+1) - sym.log(2)
    
    x = sym.symbols('x', real=True)
    p1 = sym.plot(log_cosh(x), (x,-1000,1000), show=False)
    p2 = sym.plot(replacement(x), (x,-1000,1000), show=False)
    p2[0].line_color = (0.9, 0, 0)
    p1.extend(p2)
    p1.show()
    p3 = sym.plot(log_cosh(x)-replacement(x), (x,-1000,1000))
    
    #The following should evaluate to 0, numerical values for comparison
    print(sym.simplify(log_cosh(x)-replacement(x)))
    print(sym.N(log_cosh(1000)))
    print(sym.N(replacement(1000)))
    print(sym.N(log_cosh(1000)-replacement(-1000)))
    
    #now numpy for comparison
    import numpy as np
    import matplotlib.pylab as plt
    
    def np_replacement(x):
        return np.sign(x)*x + np.log(np.exp(-np.sign(x)*2*x)+1) -np.log(2)
    
    np_x = np.linspace(-1000,1000,100001, endpoint=True)
    plt.plot(np_x, np_replacement(np_x))
    plt.show()

**Comment by asmeurer** [comment]
Numpy is choking trying to evaluate the expression, since the intermediate numbers are too big. 

```
>>> import numpy
>>> numpy.cosh(1000)
/Users/aaronmeurer/Documents/mypython/mypython/mypython.py:1: RuntimeWarning: overflow encountered in cosh
  # Define globals first so that names from this module don't get included
inf
```

We would need some way to evaluate the expression using mpmath. You can probably do it manually using `lambdify(x, log(cosh(x)), 'mpmath')`, but it would be nice to have a way to make plot do it automatically.

**Comment by laolux** [comment]
> Numpy is choking trying to evaluate the expression

I know, that's why I don't try to evaluate `numpy.cosh(1000)` in my example, but rather use the equivalent form of `numpy.log(numpy.cosh(1000))`, in my mwe found as `np_replacement`.
Anyhow, it's only for demonstration purposes and does this rather well given sufficiently fine discretization. And exactly that behaviour should be reproduced by sympy given my `replacement` class. The biggest number occuring during the evaluation of `replacement(1000)` is 1000. The "correction term" `sympy.log(sympy.exp(-sympy.sign(x)*2*x)+1)` as approximately 0 for large x, but becomes important for small x and eventually reaches log(2) for x=0.

Long story short, big numbers are not the issue after the reformulation of the problem, but the plot still does not look nice. And the reformulation is exact, so in principle sympy should be able to do it (probably it already can if "encouraged" by something besides sym.simplify, but I am not used enough to this great library yet to try and tell).

**Comment by asmeurer** [comment]
I'm not sure how to get SymPy to rewrite it to the given form. `log(cosh(x)).rewrite(exp)` works, but you then need to apply some other transformations to get it in that form.  Also, I suspect that the transformation is only correct for real x.

More generally automatically rewriting expressions to more numerically stable forms is something we'd like to be able to do, but it's a very difficult problem.

**Comment by laolux** [comment]
Yes, the transformation only does work for real x, but I supplied sympy with that information.

Using your hint to use `log(cosh(x)).rewrite(exp)` I almost got sympy to simplify everything, I only had to substitute for real posive y before before using `simplify`. Strangely enough it did not work using y directly or using a negative real z instead. But I suppose this should be discussed elsewhere, as I originally opened the ticket because plotting did not look as great as usual and it specifically asked me to report the plotting issue as a bug.

**Comment by abcde2342** [comment]
I fixed it by using Taylor series. But this issue seems to have been fixed by sympy. Plase take a look at my patch.

```
from __future__ import annotations

import math

from typing import Any

import mpmath as mp

import numpy as np

mp.mp.dps = 100

CASE_TITLE = 'handling log(cosh(x)) for large x'

CASE_FUNCTION = 'log(cosh(x))'

CASE_ID = 'sympy/sympy#12671'

FINITE_DIFF_FIXED_STEP = mp.mpf("1e-5")

def relerr(v, ref):
    is_complex = isinstance(v, complex) or isinstance(ref, (complex, mp.ctx_mp_python.mpc))
    v = mp.mpc(v) if is_complex else mp.mpf(v)
    ref = mp.mpc(ref) if is_complex else mp.mpf(ref)
    if abs(ref) == 0:
        return float(abs(v - ref))
    return float(abs((v - ref) / ref))

def orders(before, after):
    if not math.isfinite(before) or before <= 0:
        return 'inf' if after == 0 or math.isfinite(after) else 'nan'
    if after == 0:
        return 'inf'
    if not math.isfinite(after):
        return '-inf'
    return f'{math.log10(before / after):.2f}'

def log1p_taylor_scalar(s, terms=8):
    total = 0.0
    power = 1.0
    for n in range(1, terms + 1):
        power *= s
        total += (-1.0) ** (n + 1) * power / n
    return total

def log_cosh_naive(x):
    with np.errstate(over='ignore', invalid='ignore'):
        return float(np.log(np.cosh(float(x))))

def log_cosh_taylor(x, terms=8):
    a = abs(float(x))
    s = math.exp(-2.0 * a)
    return a - math.log(2.0) + log1p_taylor_scalar(s, terms=terms)

def _mp_value(x):
    if isinstance(x, (complex, mp.ctx_mp_python.mpc)):
        return mp.mpc(x)
    return mp.mpf(str(float(x))) if isinstance(x, np.generic) else mp.mpf(str(x))

def _finite_difference_step(x):
    return FINITE_DIFF_FIXED_STEP

def _finite_difference_derivative(f, x):
    x = _mp_value(x)
    fx = f(x)
    h = _finite_difference_step(x)

    def candidate(step):
        return ((f(x + step) - fx) / step, step)
    for step in (h, -h):
        try:
            d, used = candidate(step)
            if mp.isfinite(abs(d)):
                return (d, used, fx)
        except Exception:
            continue
    return (mp.nan, h, fx)

def _condition_from_finite_difference(f, x):
    d, h, fx = _finite_difference_derivative(f, x)
    x = _mp_value(x)
    if abs(fx) == 0:
        value = abs(d)
    else:
        value = abs(x * d / fx)
    return (float(value), h)

def _format_value(value: Any) -> str:
    try:
        if isinstance(value, (mp.ctx_mp_python.mpf, mp.ctx_mp_python.mpc)):
            return mp.nstr(value, 17)
        if isinstance(value, complex):
            return str(value)
        if isinstance(value, np.generic):
            return repr(value.item())
        return repr(value)
    except Exception:
        return str(value)

def case_sympy_sympy_12671() -> dict[str, Any]:
    """Compute and return reviewer evidence for sympy/sympy#12671."""
    case = 'sympy/sympy#12671'
    expressions = ('log(cosh(x))', 'mpmath log(cosh(x))', 'abs(x)-log(2)+(s - s**2/2 + s**3/3 - ...), s=exp(-2*abs(x))')
    xval = 1000.0
    ref = mp.log(mp.cosh(mp.mpf(str(xval))))
    before_raw = log_cosh_naive(xval)
    before = relerr(before_raw, ref) if math.isfinite(before_raw) else math.inf
    taylor_value = log_cosh_taylor(xval)
    after = relerr(taylor_value, ref)
    condition = _condition_from_finite_difference(lambda x: mp.log(mp.cosh(x)), xval)
    row = {'condition_number': condition[0], 'original_expression': expressions[0], 'oracle_expression': expressions[1], 'taylor_expression': expressions[2], 'original_value': _format_value(before_raw), 'oracle_value': _format_value(ref), 'taylor_value': _format_value(taylor_value), 'original_relative_error': before, 'taylor_relative_error': after, 'taylor_orders_improved': orders(before, after)}
    return ([row], expressions[2])

def print_case_report() -> None:
    rows, patch_idea = case_sympy_sympy_12671()
    print(f"{CASE_ID}: {CASE_TITLE}")
    print(f"  function: {CASE_FUNCTION}")
    print(f"  patch: {patch_idea}")
    for row in rows:
        print("  evidence:")
        print(f"    condition_number={float(row['condition_number']):.6e}")
        print(f"    original_expression={row['original_expression']}")
        print(f"    oracle_expression={row['oracle_expression']}")
        print(f"    taylor_expression={row['taylor_expression']}")
        print(f"    original_value={row['original_value']}")
        print(f"    oracle_value={row['oracle_value']}")
        print(f"    taylor_value={row['taylor_value']}")
        print(f"    original_relative_error={float(row['original_relative_error']):.6e}")
        print(f"    taylor_relative_error={float(row['taylor_relative_error']):.6e}")
        print(f"    taylor_orders_improved={row['taylor_orders_improved']}")

if __name__ == "__main__":
    print_case_report()

```
